File: backend/app/models/journal.py
from datetime import datetime
from typing import List, Optional, Literal

from pydantic import BaseModel

from .chat import Message

class JournalMetadata(BaseModel):
    """
    Metadata for a journal entry.
    """
    id: str
    filename: str
    title: str
    date: datetime
    message_count: int
    duration_seconds: Optional[int] = None
    mode: Literal["chat", "write"] = "chat"

class Journal(JournalMetadata):
    """
    Full journal entry with messages.
    """
    messages: List[Message]
    raw_content: str

class CreateJournalRequest(BaseModel):
    """
    Request to create or update a journal entry.
    """
    session_id: str
    journal_id: Optional[str] = None  # If provided, updates existing journal
    messages: List[Message]
    title: Optional[str] = None  # Auto-generate if null
    mode: Literal["chat", "write"] = "chat"

class UpdateWriteContentRequest(BaseModel):
    """
    Request to update write mode content (as a user message).
    """
    session_id: str
    journal_id: Optional[str] = None
    content: str  # The write mode content as a user message
    title: Optional[str] = None

class AskAIRequest(BaseModel):
    """
    Request to ask AI for input on write mode content.
    """
    session_id: str
    journal_id: Optional[str] = None
    content: str  # The write mode content
    conversation_history: List[Message] = []

class UpdateJournalTitleRequest(BaseModel):
    """
    Request to update journal title.
    """
    journal_id: str
    title: str

File: frontend/lib/types/journal.ts
/**
 * Journal-related TypeScript types.
 * 
 * These interfaces match the backend Pydantic models.
 */

import { Message } from './chat';

export interface JournalMetadata {
  id: string;
  filename: string;
  title: string;
  date: string;  // ISO string from backend
  message_count: number;
  duration_seconds: number | null;
  mode: "chat" | "write";
}

export interface Journal extends JournalMetadata {
  messages: Message[];
  raw_content: string;
}

export interface CreateJournalRequest {
  session_id: string;
  journal_id: string | null;
  messages: Message[];
  title: string | null;
  mode: "chat" | "write";
}

export interface UpdateWriteContentRequest {
  session_id: string;
  journal_id: string | null;
  content: string;
  title: string | null;
}

export interface AskAIRequest {
  session_id: string;
  journal_id: string | null;
  content: string;
  conversation_history: Message[];
}

export interface UpdateJournalTitleRequest {
  journal_id: string;
  title: string;
}

export interface JournalListResponse {
  journals: JournalMetadata[];
  total: number;
  limit: number;
  offset: number;
}

